### DISCUSSION TITLE
Proposal: new variant of `With` helper for improved conditional chaining

### DISCUSSION THREAD

**Discussion by jedevc** [dagger/dagger#12162] [Ideas]
Came up as an idea when discussing with @shykes. Currently, we can use this pattern:

```go
ctr = dag.
	Container().
	From("alpine").
	With(func (ctr *dagger.Container) *dagger.Container {
		if !dev {
			return ctr
		}
		return ctr.WithExec([]string{"go", "env"})
	}).
	WithExec([]string{"go", "build", "-o", "binary", "./pkg"})
```

`With` can be used to easily insert little conditionals into the chain, without breaking it! However, it's a bit unwiedly. What would be really nice would be if we could:

```go
ctr = dag.
	Container().
	From("alpine").
	IfWith(dev, func (ctr *dagger.Container) *dagger.Container {
		return ctr.WithExec([]string{"go", "env"})
	}).
	WithExec([]string{"go", "build", "-o", "binary", "./pkg"})
```

This is shorter and easier to read now (`IfWith` name TBD, it could potentially just be `If`).

cc @helderco wdyt?

**Comment by helderco** [comment]
Yeah sgtm. Just `If` is a reserved keyword in  languages that start function names with a lowercase, so I prefer `IfWith`, since it's also suggestive that it works like `With`.
